components/script/dom/idbopendbrequest.rs

```diff
@@ -0,0 +1,385 @@
+        &self,
+        name: String,
+        request_version: Option<u64>,
+        db_version: u64,
+    ) -> (Fallible<DomRoot<IDBDatabase>>, bool) {
+        // Step 5-6
+        let request_version = match request_version {
+            Some(v) => v,
+            None => {
+                if db_version == 0 {
+                    1
+                } else {
+                    db_version
+                }
+            },
+        };
+
+        // Step 7
+        if request_version < db_version {
+            return (Err(Error::Version), false);
+        }
+
+        // Step 8-9
+        let open_request = self.open_request.root();
+        let global = open_request.global();
+        let connection = IDBDatabase::new(
+            &global,
+            DOMString::from_string(name.clone()),
+            request_version,
+        );
+
+        // Step 10
+        if request_version > db_version {
+            // FIXME:(rasviitanen) Do step 10.1-10.5
+            // connection.dispatch_versionchange(db_version, Some(request_version));
+            // Step 10.6
+            open_request.upgrade_db_version(&*connection, request_version);
+            // Step 11
+            (Ok(connection), true)
+        } else {
+            // Step 11
+            (Ok(connection), false)
+        }
+    }
+
+    fn handle_delete_db(&self, result: Result<(), ()>) {
+        let open_request = self.open_request.root();
+        let global = open_request.global();
+        open_request.idbrequest.set_ready_state_done();
+
+        match result {
+            Ok(_) => {
+                let window = global.as_window();
+                let cx = window.get_cx();
+                let _ac = enter_realm(&*open_request);
+                rooted!(in(*cx) let mut answer = UndefinedValue());
+                open_request.set_result(answer.handle());
+
+                let event = Event::new(
+                    &global,
+                    Atom::from("success"),
+                    EventBubbles::DoesNotBubble,
+                    EventCancelable::NotCancelable,
+                );
+                event.upcast::<Event>().fire(open_request.upcast());
+            },
+            Err(_e) => {
+                // FIXME(rasviitanen) Set the error of request to the
+                // appropriate error
+
+                let event = Event::new(
+                    &global,
+                    Atom::from("error"),
+                    EventBubbles::Bubbles,
+                    EventCancelable::Cancelable,
+                );
+                event.upcast::<Event>().fire(open_request.upcast());
+            },
+        }
+    }
+}
+
+#[dom_struct]
+pub struct IDBOpenDBRequest {
+    idbrequest: IDBRequest,
+}
+
+impl IDBOpenDBRequest {
+    pub fn new_inherited() -> IDBOpenDBRequest {
+        IDBOpenDBRequest {
+            idbrequest: IDBRequest::new_inherited(),
+        }
+    }
+
+    pub fn new(global: &Window) -> DomRoot<IDBOpenDBRequest> {
```

Comment:
This should take a &GlobalScope argument, since these can be created from worker global scopes as well.